Added caching, prevented created ItemStacks where possible and simplified use of Streams.

metalshark opened the pull request:
Had a couple of servers-side crashes where it pointed back to this mod (part of SkyFactory One). Changed round the code to improve performance and we now use these changes. The biggest changes are caching and not creating ItemStacks for functions argument inside loops. Crash reports for prosperity:

Crash Report 1:
```
java.lang.Error: ServerHangWatchdog detected that a single server tick took 60.00 seconds (should be max 0.05)
	at net.minecraftforge.eventbus.EventBus$$Lambda$2542/173247103.invoke(Unknown Source) ~[?:?] {}
	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:302) ~[eventbus-4.0.0.jar:?] {}
	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:283) ~[eventbus-4.0.0.jar:?] {}
	at net.minecraftforge.event.ForgeEventFactory.gatherCapabilities(ForgeEventFactory.java:597) ~[forge:?] {re:mixin,pl:runtimedistcleaner:A,re:classloading,pl:runtimedistcleaner:A}
	at net.minecraftforge.event.ForgeEventFactory.gatherCapabilities(ForgeEventFactory.java:591) ~[forge:?] {re:mixin,pl:runtimedistcleaner:A,re:classloading,pl:runtimedistcleaner:A}
	at net.minecraftforge.common.capabilities.CapabilityProvider.gatherCapabilities(CapabilityProvider.java:48) ~[forge:?] {re:computing_frames,re:mixin,re:classloading}
	at net.minecraft.item.ItemStack.forgeInit(ItemStack.java:998) ~[?:?] {re:mixin,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A,re:classloading,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A}
	at net.minecraft.item.ItemStack.<init>(ItemStack.java:127) ~[?:?] {re:mixin,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A,re:classloading,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A}
	at net.minecraft.item.ItemStack.<init>(ItemStack.java:115) ~[?:?] {re:mixin,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A,re:classloading,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A}
	at net.minecraft.item.ItemStack.<init>(ItemStack.java:107) ~[?:?] {re:mixin,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A,re:classloading,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A}
	at novamachina.exnihilosequentia.common.registries.CrucibleRegistry.lambda$isMeltable$1(CrucibleRegistry.java:38) ~[exnihilosequentia:1.16-20211004011236] {re:classloading}
	at novamachina.exnihilosequentia.common.registries.CrucibleRegistry$$Lambda$15282/390906764.test(Unknown Source) ~[?:?] {}
	at java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90) ~[?:1.8.0_201] {}
	at java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1359) ~[?:1.8.0_201] {}
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126) ~[?:1.8.0_201] {}
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498) ~[?:1.8.0_201] {}
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485) ~[?:1.8.0_201] {}
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471) ~[?:1.8.0_201] {}
	at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230) ~[?:1.8.0_201] {}
	at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196) ~[?:1.8.0_201] {}
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_201] {}
	at java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:449) ~[?:1.8.0_201] {}
	at novamachina.exnihilosequentia.common.registries.CrucibleRegistry.isMeltable(CrucibleRegistry.java:38) ~[exnihilosequentia:1.16-20211004011236] {re:classloading}
	at novamachina.exnihilosequentia.common.tileentity.crucible.MeltableItemHandler.isItemValid(MeltableItemHandler.java:36) ~[exnihilosequentia:1.16-20211004011236] {re:classloading}
	at net.minecraftforge.items.ItemStackHandler.insertItem(ItemStackHandler.java:84) ~[forge:?] {re:classloading}
	at novamachina.exnihilosequentia.common.tileentity.crucible.MeltableItemHandler.insertItem(MeltableItemHandler.java:25) ~[exnihilosequentia:1.16-20211004011236] {re:classloading}
	at com.lothrazar.cyclic.base.TileEntityBase.moveItems(TileEntityBase.java:312) ~[cyclic:1.16.5-1.5.7] {re:classloading}
	at com.lothrazar.cyclic.base.TileEntityBase.moveItems(TileEntityBase.java:286) ~[cyclic:1.16.5-1.5.7] {re:classloading}
	at com.lothrazar.cyclic.block.cable.item.TileCableItem.normalFlow(TileCableItem.java:89) ~[cyclic:1.16.5-1.5.7] {re:classloading}
	at com.lothrazar.cyclic.block.cable.item.TileCableItem.func_73660_a(TileCableItem.java:69) ~[cyclic:1.16.5-1.5.7] {re:classloading}
	at net.minecraft.world.World.func_217391_K(World.java:491) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
	at net.minecraft.world.server.ServerWorld.func_72835_b(ServerWorld.java:371) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:mixins.stevekung's_lib.json:server.level.MixinServerLevel,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:851) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:globaldataandresourcepacks.mixins.json:MinecraftServerMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:291) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:787) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:globaldataandresourcepacks.mixins.json:MinecraftServerMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer.func_240802_v_(MinecraftServer.java:642) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:globaldataandresourcepacks.mixins.json:MinecraftServerMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer.func_240783_a_(MinecraftServer.java:232) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:globaldataandresourcepacks.mixins.json:MinecraftServerMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer$$Lambda$14313/2117013965.run(Unknown Source) ~[?:?] {}
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_201] {}
```

Crash Report 2:
```
java.lang.Error: ServerHangWatchdog detected that a single server tick took 60.02 seconds (should be max 0.05)
	at com.bloodnbonesgaming.topography.common.config.Preset.fireEventSubscribers(Preset.java:120) ~[topography:1.17.2] {re:classloading}
	at com.bloodnbonesgaming.topography.dedicated.ServerEventHandler.onAllEvents(ServerEventHandler.java:253) ~[topography:1.17.2] {re:classloading}
	at net.minecraftforge.eventbus.ASMEventHandler_694_ServerEventHandler_onAllEvents_Event.invoke(.dynamic) ~[?:?] {}
	at net.minecraftforge.eventbus.ASMEventHandler.invoke(ASMEventHandler.java:85) ~[eventbus-4.0.0.jar:?] {}
	at net.minecraftforge.eventbus.EventBus$$Lambda$2534/188747966.invoke(Unknown Source) ~[?:?] {}
	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:302) ~[eventbus-4.0.0.jar:?] {}
	at net.minecraftforge.eventbus.EventBus.post(EventBus.java:283) ~[eventbus-4.0.0.jar:?] {}
	at net.minecraftforge.event.ForgeEventFactory.gatherCapabilities(ForgeEventFactory.java:597) ~[forge:?] {re:mixin,pl:runtimedistcleaner:A,re:classloading,pl:runtimedistcleaner:A}
	at net.minecraftforge.event.ForgeEventFactory.gatherCapabilities(ForgeEventFactory.java:591) ~[forge:?] {re:mixin,pl:runtimedistcleaner:A,re:classloading,pl:runtimedistcleaner:A}
	at net.minecraftforge.common.capabilities.CapabilityProvider.gatherCapabilities(CapabilityProvider.java:48) ~[forge:?] {re:computing_frames,re:mixin,re:classloading}
	at net.minecraft.item.ItemStack.forgeInit(ItemStack.java:998) ~[?:?] {re:mixin,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A,re:classloading,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A}
	at net.minecraft.item.ItemStack.<init>(ItemStack.java:127) ~[?:?] {re:mixin,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A,re:classloading,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A}
	at net.minecraft.item.ItemStack.<init>(ItemStack.java:115) ~[?:?] {re:mixin,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A,re:classloading,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A}
	at net.minecraft.item.ItemStack.<init>(ItemStack.java:107) ~[?:?] {re:mixin,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A,re:classloading,xf:fml:placebo:placeboitemusehook,xf:fml:forge:filled_map.4,xf:fml:forge:itemstack,pl:runtimedistcleaner:A}
	at novamachina.exnihilosequentia.common.registries.HammerRegistry.findRecipe(HammerRegistry.java:35) ~[exnihilosequentia:1.16-20211117-154402] {re:classloading}
	at novamachina.exnihilosequentia.common.registries.HammerRegistry.isHammerable(HammerRegistry.java:29) ~[exnihilosequentia:1.16-20211117-154402] {re:classloading}
	at net.blay09.mods.excompressum.compat.exnihilosequentia.ExNihiloSequentiaAddon.isHammerable(ExNihiloSequentiaAddon.java:154) ~[excompressum:4.0.4] {re:classloading}
	at net.blay09.mods.excompressum.registry.ExNihilo.isHammerable(ExNihilo.java:26) ~[excompressum:4.0.4] {re:classloading}
	at net.blay09.mods.excompressum.tile.AutoHammerTileEntity.isRegistered(AutoHammerTileEntity.java:392) ~[excompressum:4.0.4] {re:classloading}
	at net.blay09.mods.excompressum.tile.AutoHammerTileEntity$2.isItemValid(AutoHammerTileEntity.java:66) ~[excompressum:4.0.4] {re:classloading}
	at net.minecraftforge.items.ItemStackHandler.insertItem(ItemStackHandler.java:84) ~[forge:?] {re:classloading}
	at net.blay09.mods.excompressum.utils.DefaultItemHandler.insertItem(DefaultItemHandler.java:27) ~[excompressum:4.0.4] {re:classloading}
	at net.blay09.mods.excompressum.utils.ItemHandlerAutomation.insertItem(ItemHandlerAutomation.java:34) ~[excompressum:4.0.4] {re:classloading}
	at com.lothrazar.cyclic.base.TileEntityBase.moveItems(TileEntityBase.java:312) ~[cyclic:1.16.5-1.5.7] {re:classloading}
	at com.lothrazar.cyclic.base.TileEntityBase.moveItems(TileEntityBase.java:286) ~[cyclic:1.16.5-1.5.7] {re:classloading}
	at com.lothrazar.cyclic.block.cable.item.TileCableItem.normalFlow(TileCableItem.java:89) ~[cyclic:1.16.5-1.5.7] {re:classloading}
	at com.lothrazar.cyclic.block.cable.item.TileCableItem.func_73660_a(TileCableItem.java:69) ~[cyclic:1.16.5-1.5.7] {re:classloading}
	at net.minecraft.world.World.func_217391_K(World.java:491) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:runtimedistcleaner:A}
	at net.minecraft.world.server.ServerWorld.func_72835_b(ServerWorld.java:371) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:mixins.stevekung's_lib.json:server.level.MixinServerLevel,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:851) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:globaldataandresourcepacks.mixins.json:MinecraftServerMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:291) ~[?:?] {re:classloading,pl:accesstransformer:B}
	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:787) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:globaldataandresourcepacks.mixins.json:MinecraftServerMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer.func_240802_v_(MinecraftServer.java:642) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:globaldataandresourcepacks.mixins.json:MinecraftServerMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer.func_240783_a_(MinecraftServer.java:232) ~[?:?] {re:mixin,pl:accesstransformer:B,pl:runtimedistcleaner:A,re:classloading,pl:accesstransformer:B,pl:mixin:APP:globaldataandresourcepacks.mixins.json:MinecraftServerMixin,pl:mixin:A,pl:runtimedistcleaner:A}
	at net.minecraft.server.MinecraftServer$$Lambda$14301/228840276.run(Unknown Source) ~[?:?] {}
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_201] {}```

metalshark committed:
Added caching, prevented created ItemStacks where possible and simplified use of Streams.

metalshark commented on src/main/java/novamachina/exnihilosequentia/common/registries/CrookRegistry.java:
```diff
@@ -10,33 +10,32 @@
 
-import java.util.ArrayList;
-import java.util.List;
+import java.util.*;
```
```
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
```

If not happy with the wildcard.

NovaMachina replied:
As per the contributing guidelines:
> No bulk imports. These are considered bad practice and should not be done. Instead, list out all the individual imports for the class.

metalshark commented on src/main/java/novamachina/exnihilosequentia/common/registries/SieveRegistry.java:
```diff
@@ -1,26 +1,28 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.Ingredient;
 import net.minecraft.util.IItemProvider;
+import novamachina.exnihilosequentia.common.utility.IngredientUtils;
```
Accidental re-arrangement of imports.

metalshark commented on src/main/java/novamachina/exnihilosequentia/api/registry/ICrucibleRegistry.java:
```diff
@@ -9,9 +11,17 @@
 
     CrucibleRecipe findRecipe(IItemProvider item);
 
+    CrucibleRecipe findRecipeByItemStack(ItemStack itemStack);
```
Added for convenience, can be re-worked without changing the implementation. Deliberately given different names to avoid lookup overhead associated with overloading the same function with different call signatures.

NovaMachina reviewed (changes requested):
This looks pretty good. However, I do have a couple of questions:
1. Does the caching at all impact the player's ability to reload datapacks while in game?
2. In another issue someone pointed out that the usage Java Streams was negatively impacting performance and provided values to show this. Have you experienced any issues related to performance using streams rather than for loops? What is your rational for switching to streams?
3. Did you look into using parallelized streams? I could see this reducing the performance overhead of sequential streams.

metalshark commented:
1. In some cases I’m caching negative responses, this may impact dynamic
loading and can certainly be changed to only cache positive results.

2. Streams perform worse than loops in all JVMs I’m aware of, about 2-10x
worse performance on testing I did last year after hearing about this. To
be clear this is 2-10x performance for the iteration part. Moving to using
caching reduced the CPU usage of the mod from over 50% to less than 0.5%
(most due to ItemStack creation being seemingly so costly). So personally
am in favour of the readability and maintainability of streams as it
doesn’t appear to be impacting overall performance as much as the actions
performed per iteration. It’s trivial to rewrite into loops if you’d prefer
every last drop of performance though, but will let you weight up the
differences.

Understood with bulk imports, left a comment as the import order was
unnecessarily switched around too in one file.

Will submit an update addressing the points but leaving in the use of
streams unless you feel otherwise.

On Tue, 23 Nov 2021 at 00:19, Jacob Williams ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> This looks pretty good. However, I do have a couple of questions:
>
>    1. Does the caching at all impact the player's ability to reload
>    datapacks while in game?
>    2. In another issue someone pointed out that the usage Java Streams
>    was negatively impacting performance and provided values to show this. Have
>    you experienced any issues related to performance using streams rather than
>    for loops? What is your rational for switching to streams?
>
> ------------------------------
>
> In
> src/main/java/novamachina/exnihilosequentia/common/registries/CrookRegistry.java
> <https://github.com/NovaMachina-Mods/ExNihiloSequentia/pull/282#discussion_r754725526>
> :
>
> > @@ -10,33 +10,32 @@
>  import novamachina.exnihilosequentia.common.utility.ExNihiloLogger;
>  import org.apache.logging.log4j.LogManager;
>
> -import java.util.ArrayList;
> -import java.util.List;
> +import java.util.*;
>
> As per the contributing guidelines:
>
> No bulk imports. These are considered bad practice and should not be done.
> Instead, list out all the individual imports for the class.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NovaMachina-Mods/ExNihiloSequentia/pull/282#pullrequestreview-813134294>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABD2FGW65A4RYKZJOLYHMTUNLM2HANCNFSM5IRSGQLQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

metalshark committed:
Fixed coding nits for imports.

metalshark committed:
When caching negative outcomes of non-existant recipes or recipe outcomes in registries, clear cache when setting recipes to allow for dynamic loading.

metalshark commented:
Fixed the coding nits for imports.

Looking again at the caching it's all based on recipes and is cleared if the recipes are cleared, however am now clearing the cache (if storing negative results) if setRecipes is called to add recipes dynamically. This doesn't apply to CrucibleRegistry because it is returning a null (which is never cached). 

Beyond streams vs loops there are two other potentionally contentious issues:
1. Moving the listing of unique ingredients at the beginning of SieveRegistry.getRecipeList into a separate function and potentionally caching it. Testing showed performance is negligable and am not really bothered about it staying inline (which is faster than in it's own function). Just thought it was worth highlighting: https://github.com/NovaMachina-Mods/ExNihiloSequentia/pull/282/commits/aaa6d8641d4005d34d67196afce1534df5488c57#diff-6e6908bd4604969a22d6b6b810060ec5009ba10a14a40b27c4693ed3c38afc19R95
2. Dynamically enabling ores could invalidate the item drop cache of the SieveRegistry if there were matching recipes. This cache is a significant performance increase and would frown upon dynamically enabling ores due to world gen issues. This could be enabled between server shutdown/start though safely without issue: https://github.com/NovaMachina-Mods/ExNihiloSequentia/pull/282/commits/aaa6d8641d4005d34d67196afce1534df5488c57#diff-6e6908bd4604969a22d6b6b810060ec5009ba10a14a40b27c4693ed3c38afc19R61

NovaMachina commented:
> 2\. Dynamically enabling ores could invalidate the item drop cache of the SieveRegistry if there were matching recipes. This cache is a significant performance increase and would frown upon dynamically enabling ores due to world gen issues. This could be enabled between server shutdown/start though safely without issue: [aaa6d86#diff-6e6908bd4604969a22d6b6b810060ec5009ba10a14a40b27c4693ed3c38afc19R61](https://github.com/NovaMachina-Mods/ExNihiloSequentia/commit/aaa6d8641d4005d34d67196afce1534df5488c57#diff-6e6908bd4604969a22d6b6b810060ec5009ba10a14a40b27c4693ed3c38afc19R61)

Enabling ores only happens when a configuration file changes, which I believe requires a restart of the game, or whenever an addon enables the ores, which happens before this cache would even be created.

Everything looks good otherwise!
